### PR TITLE
MINOR: Caching layer should forward record timestamp

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
\cc @guozhangwang @bbejeck

**Review by guozhangwang — COMMENTED** [review]
Add a unit test for this case?

**Comment by mjsax** [comment]
@guozhangwang Done. Extended existing unit test

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 1]
```diff
@@ -136,24 +135,9 @@ public void before() throws InterruptedException {
...
-        };
-        aggregator = new Aggregator<String, String, Integer>() {
-            @Override
-            public Integer apply(final String aggKey, final String value, final Integer aggregate) {
-                return aggregate + value.length();
-            }
-        };
+        reducer = (value1, value2) -> value1 + ":" + value2;
+        initializer = () -> 0;
+        aggregator = (aggKey, value, aggregate) -> aggregate + value.length();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:140]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 2]
```diff
@@ -181,12 +165,7 @@ public void shouldReduce() throws Exception {
             new StringDeserializer(),
             10);
 
-        Collections.sort(results, new Comparator<KeyValue<String, String>>() {
-            @Override
-            public int compare(final KeyValue<String, String> o1, final KeyValue<String, String> o2) {
-                return KStreamAggregationIntegrationTest.compare(o1, o2);
-            }
-        });
+        Collections.sort(results, KStreamAggregationIntegrationTest::compare);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:168]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 3]
```diff
@@ -228,34 +207,28 @@ public void shouldReduceWindowed() throws Exception {
...
             comparator =
-            new Comparator<KeyValue<Windowed<String>, String>>() {
-                @Override
-                public int compare(final KeyValue<Windowed<String>, String> o1,
-                                   final KeyValue<Windowed<String>, String> o2) {
-                    final int keyComparison = o1.key.key().compareTo(o2.key.key());
-                    return keyComparison == 0 ? o1.value.compareTo(o2.value) : keyComparison;
-                }
-            };
+            Comparator.comparing((KeyValue<Windowed<String>, String> o) -> o.key.key()).thenComparing(o -> o.value);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:225]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 4]
```diff
@@ -306,12 +279,7 @@ public void shouldAggregate() throws Exception {
             new IntegerDeserializer(),
             10);
 
-        Collections.sort(results, new Comparator<KeyValue<String, Integer>>() {
-            @Override
-            public int compare(final KeyValue<String, Integer> o1, final KeyValue<String, Integer> o2) {
-                return KStreamAggregationIntegrationTest.compare(o1, o2);
-            }
-        });
+        Collections.sort(results, KStreamAggregationIntegrationTest::compare);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:282]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 5]
```diff
@@ -336,75 +304,68 @@ public void shouldAggregateWindowed() throws Exception {
...
-                new TimeWindowedDeserializer<String>(),
-                new IntegerDeserializer(),
-                String.class,
-                15);
+        final String resultFromConsoleConsumer = readWindowedKeyedMessagesViaConsoleConsumer(
+            new TimeWindowedDeserializer<String>(),
+            new IntegerDeserializer(),
+            String.class,
+            15,
+            true);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:331]: add timestamps here

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 6]
```diff
@@ -336,75 +304,68 @@ public void shouldAggregateWindowed() throws Exception {
...
             comparator =
-            new Comparator<KeyValue<Windowed<String>, Integer>>() {
-                @Override
-                public int compare(final KeyValue<Windowed<String>, Integer> o1,
-                                   final KeyValue<Windowed<String>, Integer> o2) {
-                    final int keyComparison = o1.key.key().compareTo(o2.key.key());
-                    return keyComparison == 0 ? o1.value.compareTo(o2.value) : keyComparison;
-                }
-            };
+            Comparator.comparing((KeyValue<Windowed<String>, KeyValue<Integer, Long>> o) -> o.key.key()).thenComparingInt(o -> o.value.key);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:335]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 7]
```diff
@@ -419,12 +380,7 @@ private void shouldCountHelper() throws Exception {
             new StringDeserializer(),
             new LongDeserializer(),
             10);
-        Collections.sort(results, new Comparator<KeyValue<String, Long>>() {
-            @Override
-            public int compare(final KeyValue<String, Long> o1, final KeyValue<String, Long> o2) {
-                return KStreamAggregationIntegrationTest.compare(o1, o2);
-            }
-        });
+        Collections.sort(results, KStreamAggregationIntegrationTest::compare);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:383]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 8]
```diff
@@ -471,25 +427,15 @@ public void shouldGroupByKey() throws Exception {
         stream.groupByKey(Serialized.with(Serdes.Integer(), Serdes.String()))
                 .windowedBy(TimeWindows.of(500L))
                 .count()
-                .toStream(new KeyValueMapper<Windowed<Integer>, Long, String>() {
-                    @Override
-                    public String apply(final Windowed<Integer> windowedKey, final Long value) {
-                        return windowedKey.key() + "@" + windowedKey.window().start();
-                    }
-                }).to(outputTopic, Produced.with(Serdes.String(), Serdes.Long()));
+                .toStream((windowedKey, value) -> windowedKey.key() + "@" + windowedKey.window().start()).to(outputTopic, Produced.with(Serdes.String(), Serdes.Long()));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:430]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 9]
```diff
@@ -471,25 +427,15 @@ public void shouldGroupByKey() throws Exception {
...
             new StringDeserializer(),
             new LongDeserializer(),
             10);
-        Collections.sort(results, new Comparator<KeyValue<String, Long>>() {
-            @Override
-            public int compare(final KeyValue<String, Long> o1, final KeyValue<String, Long> o2) {
-                return KStreamAggregationIntegrationTest.compare(o1, o2);
-            }
-        });
+        Collections.sort(results, KStreamAggregationIntegrationTest::compare);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:438]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 10]
```diff
@@ -568,31 +514,42 @@ public void shouldCountSessionWindows() throws Exception {
                         new Properties()),
                 t4);
 
-        final Map<Windowed<String>, Long> results = new HashMap<>();
+        final Map<Windowed<String>, KeyValue<Long, Long>> results = new HashMap<>();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:517]: add timestamps here

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 11]
```diff
@@ -662,25 +619,17 @@ public void shouldReduceSessionWindows() throws Exception {
         builder.stream(userSessionsStream, Consumed.with(Serdes.String(), Serdes.String()))
                 .groupByKey(Serialized.with(Serdes.String(), Serdes.String()))
                 .windowedBy(SessionWindows.with(sessionGap).until(maintainMillis))
-                .reduce(new Reducer<String>() {
-                    @Override
-                    public String apply(final String value1, final String value2) {
-                        return value1 + ":" + value2;
-                    }
-                }, Materialized.<String, String, SessionStore<Bytes, byte[]>>as(userSessionsStore))
+                .reduce((value1, value2) -> value1 + ":" + value2, Materialized.as(userSessionsStore))
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:622]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java [hunk 12]
```diff
@@ -662,25 +619,17 @@ public void shouldReduceSessionWindows() throws Exception {
...
                 .toStream()
-                .foreach(new ForeachAction<Windowed<String>, String>() {
-                    @Override
-                    public void apply(final Windowed<String> key, final String value) {
-                        results.put(key, value);
-                        latch.countDown();
-                    }
+                .foreach((key, value) -> {
+                    results.put(key, value);
+                    latch.countDown();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:626]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 13]
```diff
@@ -310,14 +310,39 @@ public static void waitForCompletion(final KafkaStreams streams,
...
         try (final Consumer<K, V> consumer = createConsumer(consumerConfig)) {
-            final TestCondition valuesRead = new TestCondition() {
-                @Override
-                public boolean conditionMet() {
-                    final List<KeyValue<K, V>> readData =
-                        readKeyValues(topic, consumer, waitTime, expectedNumRecords);
-                    accumData.addAll(readData);
-                    return accumData.size() >= expectedNumRecords;
-                }
+            final TestCondition valuesRead = () -> {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:313]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 14]
```diff
@@ -331,14 +356,11 @@ public boolean conditionMet() {
...
         final List<ConsumerRecord<K, V>> accumData = new ArrayList<>();
         try (final Consumer<K, V> consumer = createConsumer(consumerConfig)) {
-            final TestCondition valuesRead = new TestCondition() {
-                @Override
-                public boolean conditionMet() {
-                    final List<ConsumerRecord<K, V>> readData =
-                        readRecords(topic, consumer, waitTime, expectedNumRecords);
-                    accumData.addAll(readData);
-                    return accumData.size() >= expectedNumRecords;
-                }
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:341]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 15]
```diff
@@ -369,14 +391,11 @@ public boolean conditionMet() {
...
         try (final Consumer<Object, V> consumer = createConsumer(consumerConfig)) {
-            final TestCondition valuesRead = new TestCondition() {
-                @Override
-                public boolean conditionMet() {
-                    final List<V> readData =
-                        readValues(topic, consumer, waitTime, expectedNumRecords);
-                    accumData.addAll(readData);
-                    return accumData.size() >= expectedNumRecords;
-                }
+            final TestCondition valuesRead = () -> {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:394]: Java8 rewrite only

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 16]
```diff
@@ -401,23 +420,20 @@ public static void waitUntilMetadataIsPropagated(final List<KafkaServer> servers
...
-                    final Option<UpdateMetadataRequest.PartitionState> partitionInfo =
-                            metadataCache.getPartitionInfo(topic, partition);
-                    if (partitionInfo.isEmpty()) {
-                        return false;
-                    }
-                    final UpdateMetadataRequest.PartitionState metadataPartitionState = partitionInfo.get();
-                    if (!Request.isValidBrokerId(metadataPartitionState.basePartitionState.leader)) {
-                        return false;
-                    }
+        TestUtils.waitForCondition(() -> {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:423]: Java8 rewrite only

**Review by mjsax — COMMENTED** [review]
Contains couple of Java8 rewrite and code cleanup (tried to mark accordingly with comments)

**Review by guozhangwang — APPROVED** [review]

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.0`. PR for 1.1: #5426

**Pull request merged**